### ISSUE TITLE
Shiny 1.4.0 : broken updateDateRangeInput in some case

### ISSUE THREAD

**Issue by bthieurmel** [rstudio/shiny#2703]
# System details

Browser Version: all (Edge, Mozilla, Chrome)

Output of `sessionInfo()`:

```
R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=French_France.1252  LC_CTYPE=French_France.1252    LC_MONETARY=French_France.1252
[4] LC_NUMERIC=C                   LC_TIME=French_France.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.4.0.9000

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2      digest_0.6.20   withr_2.1.2     later_1.0.0     mime_0.7        R6_2.4.0       
 [7] xtable_1.8-4    jsonlite_1.6    magrittr_1.5    rlang_0.4.0     promises_1.1.0  tools_3.6.1    
[13] httpuv_1.5.2    yaml_2.2.0      fastmap_1.0.0   compiler_3.6.1  htmltools_0.4.0
```
Same with ```1.4.0``

### Example application *or* steps to reproduce the problem

Just use the ``updateDateRangeInput`` package example : 

```R
require(shiny)
ui <- fluidPage(
  sliderInput("n", "Day of month", 1, 30, 10),
  dateRangeInput("inDateRange", "Input date range")
)

server <- function(input, output, session) {
  observe({
    date <- as.Date(paste0("2013-04-", input$n))
    
    updateDateRangeInput(session, "inDateRange",
                         label = paste("Date range label", input$n),
                         start = date - 1,
                         end = date + 1,
                         min = date - 5,
                         max = date + 5
    )
  })
}

shinyApp(ui, server)
```

### Describe the problem in detail

In some case, js ``error`` occured and blocked shiny apps. For example settings slider to 16 : 

![image](https://user-images.githubusercontent.com/8750326/68939740-9f2ee080-07a1-11ea-8188-9c6edacc585a.png)

Or to 22 after re-init : 

![image](https://user-images.githubusercontent.com/8750326/68939808-ca193480-07a1-11ea-8ee8-64862a2ac033.png)

The error appears if you jump from 10 to 16 or from 10 to 22, bug not if you use all slider valuer from 10 to 22 for example...!

**Comment by LukasK13** [comment]
I am experiencing the same issue.
In order to track the issue, I simplified the example to use `dateInput` instead of `dateRangeInput` and only a button to trigger the bug. Note that setting `value` in `updateDateInput` is not necessary but visualizes the behavior of the bug in the app and not only in the console.
```
library(shiny)

ui <- fluidPage(
  actionButton("trigger", "Trigger Bug"),
  dateInput("inDateRange", "Input date range", value = as.Date("2013-04-05"))
)

server <- function(input, output, session) {
  observeEvent(input$trigger, {
    date <- as.Date("2013-04-20")
    
    updateDateInput(session, "inDateRange",
                         label = paste("Date range label", input$n),
                         value = date,
                         min = date - 5,
                         max = date + 5
    )
  })
}

shinyApp(ui, server)
```

Google Chrome shows the following error in the JavaScript console:
![DateInput](https://user-images.githubusercontent.com/24586740/75992846-e11e9680-5ef8-11ea-9c50-eb444fbefcbf.PNG)

I was able to track the error down:
When the min and max limits of the dateInput get updated and the new min value (here: 2013-04-15) is in the future of the current selected value (here: 2013-04-05), the following happens in `input_binding_date.js`:

1. The new minimum value is applied and the currently selected date is requested to be reselected which doesn't happen as the currently selected date (here: 2013-04-05) is before the newly set minimum limit (here: 2013-04-15). This means, that the selected date remains empty
2. The new maximum value is applied and tries to set the currently selected date (which is `null` due to the previous step). Therefore, accessing the property `toDateString` of the current date object `curValue` in the following line fails:https://github.com/rstudio/shiny/blob/01491cc6969e12ad1754383d151a85c3f87639b7/srcjs/input_binding_date.js#L199

I am going to try to send a pull request to fix this issue.
